app/setup/index.js

```diff
@@ -0,0 +1,259 @@
+            },
+            body: JSON.stringify({
+                username
+            })
+        }).then(res => res.json().then(json => {
+            if (res.status !== 200) {
+                this.setState({status: 'failure', errorMessage: json.error.message});
+            }
+            this.setState({loggedIn: json.hasAuth});
+        }));
+        // TODO: More error handling here? Probably need a catch here.
+    }
+
+    createCertificates() {
+        const {username} = this.state;
+        fetch(`${baseUrl()}/hasauth`, {
+            method: 'POST',
+            headers: {
+                'Accept': 'application/json',
+                'Content-Type': 'application/json'
+            },
+            body: JSON.stringify({
+                username
+            })
+        }).then(res => res.json().then(json => {
+            if (res.status !== 200) {
+                this.setState({status: 'failure', errorMessage: json.error.message});
+            }
+            this.setState({loggedIn: json.hasAuth});
+        }));
+    }
+
+    setTimeoutAuthDone () {
+        const {username} = this.state;
+        const checkAuth = setInterval(() => {
+            // TODO: This is not very clear for a message. Show them a link to the
+            // maybe in case they closed it or want to try again?
+            this.setState({
+                errorMessage: `We\'re waiting for authorization of [${username}].`});
+            this.verifyAuthDone();
+            if (this.state.loggedIn) {
+                // TODO: Is there a better way? On premise has the `external-data-connector`
+                // route so cant use window.location.origin.
+                clearInterval(checkAuth);
+                const appRoute = baseUrl().replace('/setup', '/');
+                window.location.assign(appRoute);
+            }
+        }, 1000);
+    }
+
+    authenticateUser () {
+        if (!this.state.username) {
+            this.setState({
+                status: 'failure',
+                errorMessage: 'Plotly username can not be empty.'
+            });
+            return;
+        }
+        if (!this.state.domain && this.state.serverType === ONPREM) {
+            this.setState({
+                status: 'failure',
+                errorMessage: 'Plotly On Premise domain can not be empty.'
+            });
+            return;
+        }
+        this.setState({errorMessage: ''});
+        this.setTimeoutAuthDone();
+        if (isWebBrowser) {
+            window.open(this.buildOauthUrl(), '_blank');
+        } else {
+            dynamicRequireElectron().shell.openExternal(this.buildOauthUrl());
+        }
+    }
+
+    updateStateWithEvent(e) {
+        this.setState({[e.target.name]: e.target.value});
+    }
+
+    render() {
+        const {errorMessage, status} = this.state;
+        const submitButton = (value) => {
+            return (
+                <button
+                    onClick={(e) => this.chooseServerType(e.target.value)}
```

Comment:
I think it's clearer to just do `onClick={() => this.setState({serverType: value})` - there is less hunting around to do (what's `e.target.value`? what does `chooseServerType` do?)